pyro/mesh/patch.py

```diff
@@ -186,6 +189,133 @@ def __eq__(self, other):
+
+        self.coord_type = 0
+
+        # Length of the side in x- and y-direction
+
+        self.Lx = ArrayIndexer(np.full((self.qx, self.qy), self.dx),
+                                 grid=self)
+        self.Ly = ArrayIndexer(np.full((self.qx, self.qy), self.dy),
+                                 grid=self)
+
+        # This is area of the side that is perpendicular to x.
+
+        self.Ax_l = self.Ly.copy()
+        self.Ax_r = self.Ly.copy()
+
+        # This is area of the side that is perpendicular to y.
+
+        self.Ay_l = self.Lx.copy()
+        self.Ay_r = self.Lx.copy()
+
+        # Volume of the cell.
+
+        self.V = ArrayIndexer(np.full((self.qx, self.qy), self.dx * self.dy),
+                              grid=self)
+
+    def __str__(self):
+        """ print out some basic information about the grid object """
+        return f"Cartesian 2D Grid: xmin = {self.xmin}, xmax = {self.xmax}, " + \
+               f"ymin = {self.ymin}, ymax = {self.ymax}, " + \
+               f"nx = {self.nx}, ny = {self.ny}, ng = {self.ng}"
+
+
+class SphericalPolar(Grid2d):
+    """
+    This class defines a spherical polar grid.
+    This is technically a 2D geometry but assumes azimuthal symmetry.
+
+    Define:
+    r = x
+    theta = y
+    """
+
+    def __init__(self, nx, ny, ng=1,
+                 xmin=0.2, xmax=1.0, ymin=0.0, ymax=1.0):
+
+        # Make sure theta is within [0, PI]
+        assert ymin >= 0.0 and ymax <= np.pi, "y or \u03b8 should be within [0, \u03c0]."
+
+        # Make sure the ghost cells doesn't extend out negative x(r)
+        assert xmin - ng*(xmax-xmin)/nx >= 0.0, \
+            "xmin (r-direction), must be large enough so ghost cell doesn't have negative x."
+
+        super().__init__(nx, ny, ng, xmin, xmax, ymin, ymax)
+
+        self.coord_type = 1
+
+        # Length of the side along r-direction, dr
+
+        self.Lx = ArrayIndexer(np.full((self.qx, self.qy), self.dx),
+                               grid=self)
+
+        # Length of the side along theta-direction, r*dtheta
+
+        self.Ly = ArrayIndexer(np.full((self.qx, self.qy), self.x2d*self.dy),
+                               grid=self)
+
+        # Returns an array of the face area that points in the r(x) direction.
+        # dL_theta x dL_phi = r^2 * sin(theta) * dtheta * dphi
+
+        # dAr_l = - r{i-1/2}^2 * 2pi * cos(theta{i+1/2}) - cos(theta{i-1/2})
+        area_x_l = np.abs(-2.0 * np.pi * self.xl2d.v(buf=self.ng)**2 *
+                 (np.cos(self.yr2d.v(buf=self.ng)) -
+                  np.cos(self.yl2d.v(buf=self.ng))))
+        self.Ax_l = ArrayIndexer(area_x_l, grid=self)
+
+        # dAr_r = - r{i+1/2}^2 * 2pi * cos(theta{i+1/2}) - cos(theta{i-1/2})
+        area_x_r = np.abs(-2.0 * np.pi * self.xr2d.v(buf=self.ng)**2 *
+                 (np.cos(self.yr2d.v(buf=self.ng)) -
+                  np.cos(self.yl2d.v(buf=self.ng))))
+        self.Ax_r = ArrayIndexer(area_x_r, grid=self)
+
```

Comment:
I don't think we need both Ax_l and Ax_r, since Ax_r[i, j] = Ax_l[i+1,j]
same for Ay, and Cartesian

Reply:
Done, And I kept the unit test to test Ax_r by doing Ax.ip(1)